src/Network/Wai/Middleware/Auth/Provider.hs

```diff
@@ -0,0 +1,208 @@
+{-# LANGUAGE DeriveGeneric       #-}
+{-# LANGUAGE GADTs               #-}
+{-# LANGUAGE OverloadedStrings   #-}
+{-# LANGUAGE QuasiQuotes         #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TemplateHaskell     #-}
+module Network.Wai.Middleware.Auth.Provider
+  ( AuthProvider(..)
+  , Provider(..)
+  , ProviderUrl(..)
+  , ProviderInfo(..)
+  , AuthUser(..)
+  , UserIdentity
+  , ProviderParser
+  , Providers
+  , mkRouteRender
+  , mkProviderParser
+  , providersTemplate
+  ) where
+
+import           Blaze.ByteString.Builder      (toByteString)
+import           Control.Arrow                 (second)
+
+import           Data.Aeson
+import           Data.Aeson.TH                 (defaultOptions, deriveJSON,
+                                                fieldLabelModifier)
+import           Data.Aeson.Types              (Parser)
+import           Data.Binary                   (Binary)
+import qualified Data.ByteString               as S
+import qualified Data.ByteString.Builder       as B
+import qualified Data.HashMap.Strict           as HM
+import           Data.Int
+import           Data.Maybe                    (fromMaybe)
+import           Data.Monoid                   ((<>))
+import qualified Data.Text                     as T
+import           Data.Text.Encoding            (decodeUtf8With)
+import           Data.Text.Encoding.Error      (lenientDecode)
+import           Data.Text.Lazy.Encoding       (encodeUtf8Builder)
+import           GHC.Generics                  (Generic)
+import           Network.HTTP.Client           (Manager)
+import           Network.HTTP.Types            (Status, renderQueryText)
+import           Network.Wai                   (Request, Response)
+import           Network.Wai.Auth.Tools        (toLowerUnderscore)
+import           Text.Blaze.Html.Renderer.Text (renderHtml)
+import           Text.Hamlet                   (Render, hamlet)
+
+type UserIdentity = S.ByteString
+
+data AuthUser = AuthUser
+  { authUserIdentity :: !UserIdentity
+  , authProviderName :: !S.ByteString
+  , authLoginTime    :: !Int64
+  } deriving (Generic, Show)
+
+instance Binary AuthUser
+
+-- | Core Authentication class, that allows for extensibility of the Auth middleware.
+class AuthProvider ap where
```

Comment:
After much tinkering, I figured, that using existential quantifier was actually a good idea and it does make things simpler. For example `mkProviderParser` would become more cumbersome because it would also have to accept a creator function on how to make a `Provider` from say an `OAuth2` provider. Creating a custom provider now is very simple:
* make a custom data type `MyProvider`,
* install it into `AuthProvider` and `FromJSON`
* create a parser using `mkProviderParser (Proxy :: Proxy MyProvider)` and you are done.

So, despite that avoiding that antipattern would simplify a bit the implementation of the package itself, it would make adding custom provider a bit trickier.